File: go/changestate/machine.go
package main

import (
	"fmt"
	"strconv"

	"github.com/frame-lang/frame-demos/changestate/framelang"
)

func NewChangeState(i int) ChangeState {
	m := &changeStateStruct{}

	// Validate interfaces
	var _ ChangeState = m
	var _ ChangeState_actions = m
	m._compartment_ = NewChangeStateCompartment(ChangeStateState_S0)
	m._compartment_.StateArgs["i"] = i
	m._compartment_.StateVars["dec"] = 1

	// Initialize domain

	// Send system start event
	e := framelang.FrameEvent{Msg: ">"}
	m._mux_(&e)
	return m
}

type ChangeStateState uint

const (
	ChangeStateState_S0 ChangeStateState = iota
	ChangeStateState_S1
	ChangeStateState_Stop
)

type ChangeState interface {
}

type ChangeState_actions interface {
	print(s string)
}

type changeStateStruct struct {
	_compartment_     *ChangeStateCompartment
	_nextCompartment_ *ChangeStateCompartment
}

//====================== Multiplexer ====================//

func (m *changeStateStruct) _mux_(e *framelang.FrameEvent) {
	switch m._compartment_.State {
	case ChangeStateState_S0:
		m._ChangeStateState_S0_(e)
	case ChangeStateState_S1:
		m._ChangeStateState_S1_(e)
	case ChangeStateState_Stop:
		m._ChangeStateState_Stop_(e)
	}

	if m._nextCompartment_ != nil {
		nextCompartment := m._nextCompartment_
		m._nextCompartment_ = nil
		if nextCompartment._forwardEvent_ != nil &&
			nextCompartment._forwardEvent_.Msg == ">" {
			m._mux_(&framelang.FrameEvent{Msg: "<", Params: m._compartment_.ExitArgs, Ret: nil})
			m._compartment_ = nextCompartment
			m._mux_(nextCompartment._forwardEvent_)
		} else {
			m._do_transition_(nextCompartment)
			if nextCompartment._forwardEvent_ != nil {
				m._mux_(nextCompartment._forwardEvent_)
			}
		}
		nextCompartment._forwardEvent_ = nil
	}
}

//===================== Machine Block ===================//

func (m *changeStateStruct) _ChangeStateState_S0_(e *framelang.FrameEvent) {
	switch e.Msg {
	case ">":
		m._compartment_.StateArgs["i"] = m._compartment_.StateArgs["i"].(int) - m._compartment_.StateVars["dec"].(int)
		m.print(strconv.Itoa((m._compartment_.StateArgs["i"].(int))))
		if (m._compartment_.StateArgs["i"].(int)) == 0 {
			compartment := NewChangeStateCompartment(ChangeStateState_Stop)
			m._transition_(compartment)
			return
		}
		compartment := NewChangeStateCompartment(ChangeStateState_S1)
		compartment.EnterArgs["i"] = m._compartment_.StateArgs["i"].(int)
		m._transition_(compartment)
		return
	}
}

func (m *changeStateStruct) _ChangeStateState_S1_(e *framelang.FrameEvent) {
	switch e.Msg {
	case ">":
		compartment := NewChangeStateCompartment(ChangeStateState_S0)
		compartment.StateArgs["i"] = e.Params["i"].(int)

		compartment.StateVars["dec"] = 1

		m._transition_(compartment)
		return
	}
}

func (m *changeStateStruct) _ChangeStateState_Stop_(e *framelang.FrameEvent) {
	switch e.Msg {
	case ">":
		m.print("done")
		return
	}
}

//=============== Machinery and Mechanisms ==============//

func (m *changeStateStruct) _transition_(compartment *ChangeStateCompartment) {
	m._nextCompartment_ = compartment
}

func (m *changeStateStruct) _do_transition_(nextCompartment *ChangeStateCompartment) {
	m._mux_(&framelang.FrameEvent{Msg: "<", Params: m._compartment_.ExitArgs, Ret: nil})
	m._compartment_ = nextCompartment
	m._mux_(&framelang.FrameEvent{Msg: ">", Params: m._compartment_.EnterArgs, Ret: nil})
}

//===================== Actions Block ===================//

func (m *changeStateStruct) print(s string) {
	fmt.Println(s)
}

//********************************************************//

// Unimplemented Actions

//********************************************************/

//=============== Compartment ==============//

type ChangeStateCompartment struct {
	State          ChangeStateState
	StateArgs      map[string]interface{}
	StateVars      map[string]interface{}
	EnterArgs      map[string]interface{}
	ExitArgs       map[string]interface{}
	_forwardEvent_ *framelang.FrameEvent
}

func NewChangeStateCompartment(state ChangeStateState) *ChangeStateCompartment {
	c := &ChangeStateCompartment{State: state}
	c.StateArgs = make(map[string]interface{})
	c.StateVars = make(map[string]interface{})
	c.EnterArgs = make(map[string]interface{})
	c.ExitArgs = make(map[string]interface{})
	return c
}

File: csharp/Countdown/machine.cs
using Countdown;
#nullable disable
namespace csharp.countdown
{

class Countdown
{

    private CountdownCompartment _compartment_;
    private CountdownCompartment _nextCompartment_;


    public Countdown(int i)
    {


        // Create and intialize start state compartment.


        this._state_ = (int)CountdownState.S0;
        this._compartment_ = new CountdownCompartment(this._state_);
        this._nextCompartment_ = null;
        this._compartment_.StateArgs["i"] = i;
        this._compartment_.StateVars["dec"] = 1;


        // Send system start event
        FrameEvent frameEvent = new FrameEvent(">", null);
        this._mux_(frameEvent);

    }

    // states enum
    private enum CountdownState
    {

        S0 = 0,
        S1 = 1,
        STOP = 2
    }
    //====================== Multiplexer ====================//

    private void _mux_(FrameEvent e)
    {
        switch (this._compartment_.state)
        {
            case (int)CountdownState.S0:
                this._sS0_(e);
                break;
            case (int)CountdownState.S1:
                this._sS1_(e);
                break;
            case (int)CountdownState.STOP:
                this._sStop_(e);
                break;
        }

        if( this._nextCompartment_ != null)
        {
            var nextCompartment = this._nextCompartment_;
            this._nextCompartment_ = null;
            if (nextCompartment._forwardEvent != null &&
               nextCompartment._forwardEvent._message == ">")
            {
                this._mux_(new FrameEvent( "<", this._compartment_.ExitArgs));
                this._compartment_ = nextCompartment;
                this._mux_(nextCompartment._forwardEvent);
            }
            else
            {
                this._doTransition_(nextCompartment);
                if (nextCompartment._forwardEvent != null)
                {
                    this._mux_(nextCompartment._forwardEvent);
                }
            }
            nextCompartment._forwardEvent = null;
        }
    }


    //===================== Machine Block ===================//

    CountdownCompartment compartment;


    private void _sS0_(FrameEvent e)
    {
        if (e._message == ">")
        {
            (this._compartment_.StateArgs["i"]) = ((int)this._compartment_.StateArgs["i"]) - ((int)this._compartment_.StateVars["dec"]);
            this.print_do(((int)this._compartment_.StateArgs["i"]).ToString());
            if (((int)this._compartment_.StateArgs["i"]) == 0) {


                compartment =  new CountdownCompartment((int)CountdownState.STOP);


                this._transition_(compartment);
                return;
            }


            compartment =  new CountdownCompartment((int)CountdownState.S1);

            compartment.EnterArgs["i"]  = (int)this._compartment_.StateArgs["i"];

            this._transition_(compartment);
            return;
        }
    }

    private void _sS1_(FrameEvent e)
    {
        if (e._message == ">")
        {


            compartment =  new CountdownCompartment((int)CountdownState.S0);

            compartment.StateArgs["i"] = (int)e._parameters["i"];

            compartment.StateVars["dec"] = 1;

            this._transition_(compartment);
            return;
        }
    }

    private void _sStop_(FrameEvent e)
    {
        if (e._message == ">")
        {
            this.print_do("done");
            return;
        }
    }

    //===================== Actions Block ===================//

    public void print_do(String s)
    {
        Console.WriteLine(s);
    }

    // Unimplemented Actions



    //=============== Machinery and Mechanisms ==============//

    private int _state_;

    private void _transition_(CountdownCompartment compartment)
    {
        this._nextCompartment_ = compartment;
    }

    private void _doTransition_(CountdownCompartment nextCompartment)
    {
        this._mux_(new FrameEvent("<", this._compartment_.ExitArgs));
        this._compartment_ = nextCompartment;
        this._mux_(new FrameEvent(">", this._compartment_.EnterArgs));
    }

}

//=============== Compartment ==============//

class CountdownCompartment
{

    public int state;

    public CountdownCompartment(int state)
    {
        this.state = state;
    }

    public Dictionary<string, object> StateArgs { get; set; } = new Dictionary<string, object>();
    public Dictionary<string, object> StateVars { get; set; } = new Dictionary<string, object>();
    public Dictionary<string, object> EnterArgs { get; set; } = new Dictionary<string, object>();
    public Dictionary<string, object> ExitArgs { get; set; } = new Dictionary<string, object>();
    public FrameEvent _forwardEvent = new FrameEvent();
}


class CountdownController : Countdown
{
        public CountdownController(int i) : base(i)
        {
        }
}

}